### PR TITLE
Inferring property fields in a class context when metaclass is present

### LINKED ISSUE #940: @property members defined in metaclasses of a base class are not correctly inferred
Ref https://github.com/PyCQA/astroid/issues/927#issuecomment-817244963

Inference works on the parent class but not the child in the following example:

```python
class BaseMeta(type):
    @property
    def __members__(cls):
        return ['a', 'property']
class Parent(metaclass=BaseMeta):
    pass
class Derived(Parent):
    pass
Parent.__members__  # [<Set.set l.10 at 0x...>]
Derived.__members__  # [<Property.__members__ l.8 at 0x...>]
```

### PR DESCRIPTION
**Pull request by nelfin**
## Steps

- [x] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [x] Write a good description on what the PR does.

## Description

If we are accessing an attribute and it is found on
`type(A).__dict__` *and* it is a data descriptor, then we resolve that
descriptor. For the case of inferring a property which is accessed as a
class attribute, this equates to the property being defined on the
metaclass (which may be anywhere in the class hierarchy).

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Related Issue

Closes #940

### CODE REVIEW

File: tests/unittest_scoped_nodes.py
Comment (hippo91) [inline tests/unittest_scoped_nodes.py:1988]: I'am probably nitpicking, but should not it be `def __members__(self)`?

File: tests/unittest_scoped_nodes.py
Comment (hippo91) [inline tests/unittest_scoped_nodes.py:2034]: I'am fine with adding such a test but should we not open an issue for that very specific case?

File: astroid/brain/brain_namedtuple_enum.py
Comment (hippo91) [inline astroid/brain/brain_namedtuple_enum.py:318]: I wonder if those modifications should not be the topic of another MR? 
Is there a corresponding issue that has been opened?
I can't see a strong relationship with other modifications of this MR.

**Review by hippo91 — APPROVED** [review]
Very nice job @nelfin! The fix is clear, documented and heavily tested! :+1: 
I approved this PR but let a few comments. 
We just have to wait for @cdce8p review before merging.

File: astroid/brain/brain_namedtuple_enum.py
Comment (hippo91) [reply astroid/brain/brain_namedtuple_enum.py:318]: I answer to myself. It appears that PyCQA/pylint#4358 is the corresponding issue.

**Review by cdce8p — CHANGES_REQUESTED** [review]
Overall it looks good. Just some small comments and a merge conflict. After those are resolved, this one can be merged.

File: astroid/brain/brain_namedtuple_enum.py
Comment (nelfin) [reply astroid/brain/brain_namedtuple_enum.py:318]: @hippo91: sorry about the confusion, the `Enum.__members__` case was the original motivation for all of these changes (via https://github.com/PyCQA/astroid/pull/927#issuecomment-817138770 and above). I lost track of the issues that referenced this, I'll update the changelog and add a comment or so referencing them.

File: tests/unittest_scoped_nodes.py
Comment (nelfin) [reply tests/unittest_scoped_nodes.py:1988]: In the rest of the cases, the methods are on the metaclass and it's deliberately `cls` as a mnemonic to remind myself that they're supposed to take a class of <type> and not an instance of <class> (for which we'd use `self` by convention), e.g.
```python
class EnumMeta(type):
    def members(cls):
        print(f'{cls} of type {type(cls)}')

class Enum(metaclass=EnumMeta):
    pass

Enum.members()
# <class '__main__.Enum'> of type <class '__main__.EnumMeta'>
```

In this test, it's `cls` because I copied it from the test above and forgot to change it whoops.

File: astroid/brain/brain_namedtuple_enum.py
Comment (nelfin) [reply astroid/brain/brain_namedtuple_enum.py:318]: @cdce8p: There were existing regression tests for `Enum.__members__` in pylint which passed due to `__members__` being uninferable (prior to #927 landing). I addressed one of them in this related PR: https://github.com/PyCQA/pylint/pull/4466. I can add additional regression tests that address the property/class-context case more directly

File: tests/unittest_scoped_nodes.py
Comment (nelfin) [reply tests/unittest_scoped_nodes.py:2034]: As I was writing up a new issue, I couldn't get my tests to fail. It turned out that the reason this test failed was because I was using the function-call style `property` instead of the decorator. This case is fixed by these changes:

Given:
```python
import astroid
cls_context, obj_context = astroid.extract_node("""
class Meta(type):
    def value(cls):
        return 123

class Example(metaclass=Meta):
    @property
    def value(self):
        return 456

Example.value  #@
Example().value  #@
""")
print(next(cls_context.infer()))
print(next(obj_context.infer()))
```

it prints

```
Const.int(value=456)
Const.int(value=456)
```

on d19758108727e1fd11e343d857d060d4f80655d1, and it prints

```
Property.value(name='value',
               doc=None,
               decorators=None,
               args=<Arguments l.8 at 0x...>,
               returns=None,
               body=[])
Const.int(value=456)
```

with these changes

File: astroid/brain/brain_namedtuple_enum.py
Comment (cdce8p) [reply astroid/brain/brain_namedtuple_enum.py:318]: > I can add additional regression tests that address the property/class-context case more directly

👍🏻 I would appreciate that

File: ChangeLog
Comment (cdce8p) [inline ChangeLog:32]: https://github.com/PyCQA/pylint/issues/2626 doesn't seem to be fixed (at least not the initial example). Am I missing something?

File: tests/unittest_scoped_nodes.py
Comment (cdce8p) [reply tests/unittest_scoped_nodes.py]: Makes sense. Thanks for taking the time to explain it!

Comment (hippo91) [reply tests/unittest_scoped_nodes.py]: OK. Thanks for the explanation

File: astroid/brain/brain_namedtuple_enum.py
Comment (hippo91) [reply astroid/brain/brain_namedtuple_enum.py:318]: Me too! :wink:

File: tests/unittest_scoped_nodes.py
Comment (hippo91) [reply tests/unittest_scoped_nodes.py]: OK. Nice!

File: astroid/brain/brain_namedtuple_enum.py
Comment (nelfin) [reply astroid/brain/brain_namedtuple_enum.py:318]: @cdce8p, @hippo91: https://github.com/PyCQA/pylint/pull/4473

File: ChangeLog
Comment (nelfin) [reply ChangeLog]: You're right, I was looking at the comment in https://github.com/PyCQA/pylint/issues/2626#issuecomment-627442384

**Review by cdce8p — APPROVED** [review]
👍🏻 Thanks again

**Pull request merged**
